workout-tracker: init at 0.10.5

bhankas opened the pull request:
## Description of changes

Add https://github.com/jovandeginste/workout-tracker as package.

A rudimentary NixOS module is also included.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
maintainers: payas -> bhankas

Commit:
workout-tracker: init at v0.10.5

Commit:
nixos/workout-tracker: init

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/1016/111

leona-ya commented on nixos/modules/services/misc/workout-tracker.nix:
```diff
@@ -0,0 +1,121 @@
+  ...
+}:
+
+with lib;
```
Can you please not use `with lib` here. See #208242

leona-ya commented on nixos/modules/services/misc/workout-tracker.nix:
```diff
@@ -0,0 +1,121 @@
+        '';
+      };
+
+      settings = mkOption {
```
This option isn't used. Please remove or use it

leona-ya commented on nixos/modules/services/misc/workout-tracker.nix:
```diff
@@ -0,0 +1,121 @@
+let
+  inherit (lib) types;
+  cfg = config.services.workout-tracker;
+  exe = lib.getExe cfg.package;
```
As this attribute is just used once, just write `${lib.getExe cfg.package}` there.

leona-ya commented on nixos/modules/services/misc/workout-tracker.nix:
```diff
@@ -0,0 +1,121 @@
+
+      package = lib.mkPackageOption pkgs "workout-tracker" { };
+
+      address = mkOption {
+        type = types.str;
+        default = "localhost";
+        description = lib.mdDoc "Web interface address.";
+      };
+
+      port = mkOption {
+        type = types.port;
+        default = 8080;
+        description = lib.mdDoc "Web interface port.";
+      };
```
I think it should be possible to include these in a RFC42 compliant settings option (you can specify the attribute explicitly there with a default)

bhankas replied:
I think I made the changes as laid out in `settings-optiona.sections.md` now, but please check. I have not written a module before.

leona-ya replied:
This is not really what I meant unfortunately. I meant that you define a `settings` option (as you did) and then autogenerate the full environment for the systemd service from it. You can find one example for this here: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/paperless.nix#L30-L34

bhankas replied:
Okay, I adopted solution from a different module, but please check.

leona-ya commented on nixos/modules/services/misc/workout-tracker.nix:
```diff
@@ -0,0 +1,121 @@
+      serviceConfig = {
+        ExecStart = "${exe}";
+        WorkingDirectory = cfg.stateDir;
+        User = cfg.user;
```
Does this service work with `DynamicUser`? If yes please activate it, if not please add `users.users.${cfg.user}` and `users.groups.${cfg.group}`

bhankas replied:
Shouldn't have any issues with `DynamicUser`. I'm bit low on systemd-unit knowledge in this area and will read more.

leona-ya commented on nixos/modules/services/misc/workout-tracker.nix:
```diff
@@ -0,0 +1,121 @@
+      address = mkOption {
+        type = types.str;
+        default = "localhost";
+        description = lib.mdDoc "Web interface address.";
```
You can also remove all `lib.mdDoc` as it has no effect anymore (just use the string),

leona-ya reviewed (changes requested):
Thank your for the contribution. Mostly looked at the module for now

bhankas commented:
@leona-ya thanks a lot for the review! I've tried to address everything with proposed changes, please take a look whenever possible.

leona-ya commented on nixos/modules/services/misc/workout-tracker.nix:
```diff
@@ -0,0 +1,123 @@
+            );
+
+          options = {
+            address = lib.mkOption {
+              type = types.str;
+              default = "localhost";
+              description = "Web interface address.";
+            };
+
+            port = lib.mkOption {
+              type = types.port;
+              default = 8080;
+              description = "Web interface port.";
+            };
+
+            stateDir = lib.mkOption {
+              default = "/var/lib/workout-tracker";
+              type = types.str;
+              description = "Workout-tracker data directory.";
+            };
+          };
```
I think I gave you a a bit wrong suggestion.
The `stateDir` shouldn't be part of `services.workout-tracker.settings`. 
For `address` and `port`, there are two possible ways to handle that
1. Replace them with one `WT_BIND` option that has `localhost:8000` as a default
2. Re-move them out of `services.workout-tracker.settings`. Then instead of setting `WT_BIND` in line 107, set `cfg.settings.WT_BIND = "${cfg.settings.address}:${toString cfg.settings.port}"` somewhere in the `config` block.

I see advantages and disadvantages with both options. Feel free to decide which you think is better (I would probably choose 2., but very light preference)

bhankas replied:
Went with option 2, reworked things fair bit, used the `default` field to set it, but user can override. WDYT?

leona-ya replied:
This looks good to me! I would probably remove `WT_BIND` from the example, as a user can think that they should set it

bhankas replied:
This service is intended to be served from an always on device, something that is more likely than not to be user's non-primary, non-interactive device. I think it will help to show people via example that this is what they can do to make it available over network. Also, `options.nixos.org` renders both default and example, so it should be fine.

I will remove it if you think that is a good idea, but this is my rationale.

leona-ya replied:
But shouldn't they then change `services.workout-tracker.address` instead of `services.workout-tracker.settings.WT_BIND`

leona-ya replied:
You need also a `defaultText` for `.settings` so that the NixOS docs build

bhankas replied:
fair point, removed.

Also having it in `default` causes issues with manual generation. So now setting it via service env, which from what I understand takes precedence over what gets added to EnvironmentFile.

bhankas replied:
> You need also a `defaultText` for `.settings` so that the NixOS docs build

Its fixed, just needed to not refer to `cfg` in default text.

leona-ya reviewed (commented):
The module looks much better. I have one more suggestion and after that still need to test the module :)

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -465,3 +465,5 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 - `documentation.man.mandoc` now by default uses `MANPATH` to set the directories where mandoc will search for manual pages.
   This enables mandoc to find manual pages in Nix profiles. To set the manual search paths via the `mandoc.conf` configuration file like before, use `documentation.man.mandoc.settings.manpath` instead.
+
+- A new module `services.workout-tracker` has been added for [Workout-tracker](https://github.com/jovandeginste/workout-tracker) service.
```
Please see the comment below the next headline

bhankas replied:
Sorry I don't understand what you mean here. This entry is the last in file, which headline are you referring to?

bhankas replied:
I got it, fixed now.

SuperSandro2000 commented on nixos/modules/services/misc/workout-tracker.nix:
```diff
@@ -0,0 +1,123 @@
+
+      settings = lib.mkOption {
+        type = lib.types.submodule {
+          freeformType =
```
Condensing this into one or two lines would make the code a lot more readable

bhankas replied:
not sure what you mean by this. If you mean formatting, I'm using `nixfmt-rfc-style` which formats vertical space on its own.

bhankas replied:
Also fixed.

SuperSandro2000 commented on nixos/modules/services/misc/workout-tracker.nix:
```diff
@@ -0,0 +1,123 @@
+          options = {
+            address = lib.mkOption {
+              type = types.str;
+              default = "localhost";
```
localhost is a domain, not an address. I am confused by this option.

SuperSandro2000 commented on nixos/modules/services/misc/workout-tracker.nix:
```diff
@@ -0,0 +1,123 @@
+        WT_BIND = "${cfg.settings.address}:${toString cfg.settings.port}";
+      };
+      serviceConfig = {
+        ExecStart = "${lib.Exe cfg.package}";
```
```suggestion
        ExecStart = lib.Exe cfg.package;
```

SuperSandro2000 commented on pkgs/by-name/wo/workout-tracker/package.nix:
```diff
@@ -0,0 +1,34 @@
+
+  doCheck = false;
+
+  proxyVendor = true;
```
Why is that required?

bhankas replied:
removed. `vendorHash = null;` is required because upstream includes their own `vendor` dir.

SuperSandro2000 commented on pkgs/by-name/wo/workout-tracker/package.nix:
```diff
@@ -0,0 +1,34 @@
+
+  vendorHash = null;
+
+  doCheck = false;
```
Why not execute tests?

SuperSandro2000 commented on nixos/modules/services/misc/workout-tracker.nix:
```diff
@@ -0,0 +1,103 @@
+{
+  options = {
+    services.workout-tracker = {
+      enable = lib.mkEnableOption "A workout tracking web application for personal use (or family, friends), geared towards running and other GPX-based activities";
```
This will be prefixed with `Whether to`.

```suggestion
      enable = lib.mkEnableOption "workout tracking web application for personal use (or family, friends), geared towards running and other GPX-based activities";
```

bhankas replied:
Done

bhankas commented:
Hi @leona-ya , @SuperSandro2000 . I've added a module test and reworked some options. Looks like `WorkingDirectory` does not work with `DynamicUser`. Its still preferable to have the dynamic user, but if you think user-customizable user and group values are better, we can go with that.

ElvishJerricco commented on nixos/tests/workout-tracker.nix:
```diff
@@ -0,0 +1,29 @@
+      start_all()
+      machine.wait_for_unit("workout-tracker.service")
+      # wait for workout-tracker to fully come up
+      machine.sleep(30)
+
+      with subtest("workout-tracker service starts"):
+          machine.succeed(
+              "curl -sSfL http://localhost:8080/ > /dev/null"
+          )
```
It would be better if this was:
```suggestion
      with subtest("workout-tracker service starts"):
          machine.wait_until_succeeds(
              "curl -sSfL http://localhost:8080/ > /dev/null",
              timeout=30
          )
```

bhankas replied:
Thanks, done!

bhankas commented:
[As discussed](https://matrix.to/#/!RRerllqmbATpmbJgCn:nixos.org/$VEVI1L6KjYvwSH53AuOWDZGIfCUlehVMZ06gBaQHJWg?via=nixos.org&via=matrix.org&via=tchncs.de) with @ElvishJerricco on matrix, `DynamicUser` is preferable to explicit one, and since the `StateDirectory` for dynamic user is always under `/var/lib/` it doesn't make sense to keep it configurable. The module test now works with on-disk db, so this PR is now again ready for review.

bhankas commented:
Hi @leona-ya , @SuperSandro2000 , anything more left to do in this PR? Can this be merged?

leona-ya reviewed (approved):
Looks good to me now. Thanks!

bhankas commented:
Thanks @leona-ya !

@SuperSandro2000 if nothing else, can we merge this?

wegank commented on pkgs/by-name/wo/workout-tracker/package.nix:
```diff
@@ -0,0 +1,32 @@
+  buildGoModule,
+  fetchFromGitHub,
+  nix-update-script,
+  ...
```
```suggestion
```

bhankas replied:
@wegank pardon my insolence, but perhaps these kind of trivial automated suggestions should be part of CI, and if not, they should be made before someone has spent multiple hours on addressing thoughtful feedback made by actual humans.

Also, this comment gives no indication why this change must be made, and frankly, at this point I do not want to.

@leona-ya my simcerest apologies if not making this change is a blocker and I've wasted your time in that case, please let me know and I'll just close the PR.

wegank replied:
I'm surprised to read this at 4 am in my time zone, I just got off my bed to say that I'm not a bot, and my suggestion isn't automated. It is part of a review following [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), if you look for "Functions should list their expected arguments as precisely as possible". I'm sorry if that wasn't clear.

Anyway, I don't want you to close the PR, so I'll merge it, and you're free to make the change in a PR or not afterwards.

bhankas replied:
Hi @wegank ,

First of all, my comment was unnecessarily rude to you, it was wrong and I am sorry for it. Lesson learned, wait and think before putting finger to keyboard.

It has been a while since I've read contributing guide and this is perhaps good reminder to do it again.

Coming back, and I really mean this with respect, that I've had such tiny, most definitely automated review comments in past, although never from you. I understand it is more work, but above reason with link to contributing guide would have been much much better, as it tells me why something is being asked of me, and help understand it wasn't just somebody being pedantic.

Ultimately, as with nixfmt, I sincerely think things, in grand scheme of things, should be automated. I was very relieved to not have to live through formatting comments, and a fix like current one can be trivially performed via a linter. If you have a preferred one, can you please link me to it so I can be more careful from next time onwards?

P.S.: Thanks for merging, I'll be sure to fix this in next commit.

wegank replied:
> I understand it is more work, but above reason with link to contributing guide would have been much much better, as it tells me why something is being asked of me, and help understand it wasn't just somebody being pedantic.

I'll keep that in mind, thanks.

> If you have a preferred one, can you please link me to it so I can be more careful from next time onwards?

I don't use any Nix linters, except perhaps nil which does give some lints for my own PRs, so I can't help you there...
